src/main/java/dk/kb/util/yaml/YAML.java

```diff
@@ -668,52 +684,293 @@ public boolean containsKey(Object path) throws NullPointerException {
+     * such as foo.zoo.bar and foo.fooz.bar, but not a path such as foo.fooz.zoo.bar
+     * <p>
+     * The implementation also supports the syntax foo.**.bar, which returns all bar's that are below 'foo' in the hierarchy.
+     * This path would match all the following values: 'foo.zoo.bar', 'foo.fooz.zoo.bar' and 'foo.anything.bar'.
+     * <p>
+     * The implementation supports conditional lookups as well. This means that a query for zoo.[foo=bar].baz returns
+     * all values for baz, that are a child of zoo and has the sibling-key foo with the value bar.
+     * @param yPath a list of path elements. This list contains all parts of a specified path, which is most likely
+     *              delivered through the {@link #visit(Object path, YAML yaml)}-method.
+     * @param yaml the current place in the YAML file being traversed. This should be an instance of a Map.
+     * @param visitor used to collect values that match the given path. The only function of this visitor is to collect
+     *               values to its internal list of extracted values: {@link MultipleValuesVisitor#extractedValues}.
+     */
+    private void traverseMap(List<String> yPath, Object yaml, MultipleValuesVisitor visitor) {
+        // Quick fix cleaning entries as [foo=bar] to foo=bar.
+        removeBracketsFromPathElement(yPath);
+
+        Map<?, ?> map = (Map<?, ?>) yaml;
+        List<String> shortenedPath = removeTraversedEntry(yPath);
+
+        Matcher conditionalMatch;
+        if (!yPath.isEmpty()){
+            conditionalMatch = ARRAY_CONDITIONAL.matcher(yPath.get(0));
+        } else {
+            conditionalMatch = ARRAY_CONDITIONAL.matcher("");
+        }
+
+        if (!yPath.isEmpty() && yPath.get(0).equals("**") && yPath.size() > 1){
+            for (Map.Entry<?, ?> entry : map.entrySet()) {
+                traverse(yPath, entry.getValue(), visitor);
+                traverse(shortenedPath, entry.getValue(), visitor);
+            }
+        } else if (!yPath.isEmpty() && yPath.get(0).equals("*")) {
+            for (Map.Entry<?, ?> entry : map.entrySet()) {
+                if (yPath.size() > 1) {
+                    traverse(shortenedPath, entry.getValue(), visitor);
+                } else {
+                    visitor.visit(entry.getValue());
+                }
+            }
+        } else if (conditionalMatch.matches()) {
+            conditionalTraverse(yPath, visitor, conditionalMatch, map, shortenedPath);
+        } else {
+            for (Map.Entry<?, ?> entry : map.entrySet()) {
+                String key = entry.getKey().toString();
+                Object value = entry.getValue();
+
+                if (yPath.size() == 1){
+                    if (key.equals(yPath.get(0))){
+                        visitor.extractedValues.add(value);
+                    }
+                } else {
+                    if (key.equals(yPath.get(0))){
+                        traverse(shortenedPath, value, visitor);
```

Comment:
Sending the map entry value breaks substitution. I have added the unit test `testVisitPathSubstitutionConditional` to trigger this.

The constructor
```Java
new YAML(value, extrapolateSystemProperties, getSubstitutors())
```
will fix this if the value is a map. Maybe we should introduce a `wrapWithSubstitutors` helper method that handles the cases of map, list & value?

Reply:
I don't have this testVisitPathSubstitutionConditional test? Is it committed?